### PR TITLE
Arya game

### PR DESCRIPTION
**Pull request by aryanagarkar**

### CODE REVIEW

File: Questions/AdditionQuestion.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Questions
+{
+    public class AdditionQuestion : Question
```
Comment (nagarkar) [inline Questions/AdditionQuestion.cs:7]: Just 'Addition' will do.

File: Questions/Question.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Questions
+{
+    abstract public class Question : MonoBehaviour
```
Comment (nagarkar) [inline Questions/Question.cs:7]: Why is this a MonoBehavior?

File: Questions/Question.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        protected string difficulty;
```
Comment (nagarkar) [inline Questions/Question.cs:9]: These are readonly.

File: Questions/Question.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        protected string preferredDisplayType;
+
+        protected int answer;
```
Comment (nagarkar) [inline Questions/Question.cs:12]: Why are possibleAnswers strings, and this an int?

File: Questions/Question.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            get { return problem; }
+        }
+
+        public int Answer
+        {
+            get { return answer; }
+        }
+
+        public string[] PossibleAnswers
```
Comment (nagarkar) [inline Questions/Question.cs:32]: why is this here?

File: Questions/Question.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+                int error = 2;
+                possibleAnswers[0] = (answer - error).ToString();
+                possibleAnswers[1] = (answer + error).ToString();
+                return possibleAnswers;
+            }
+        }
+
+        protected abstract void createQuestion();
+
+        public void setProblemAndAnswer()
```
Comment (nagarkar) [inline Questions/Question.cs:46]: wtf is this? Who names a method like this? Why?

File: Questions/SubtractionQuestion.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+{
+    public class SubtractionQuestion : Question
+    {
+        public SubtractionQuestion(string difficulty, string preferredDisplayType)
+        {
+            this.difficulty = difficulty;
+            this.preferredDisplayType = preferredDisplayType;
+        }
+
+        override protected void createQuestion()
```
Comment (nagarkar) [inline Questions/SubtractionQuestion.cs:15]: Wrong implementation. breaks the readonly rules for the properties it modifies.

File: Tests/AdditionQuestionTest.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Tests
+{
+    public class AdditionQuestionTest
+    {
+        [Test]
+        public void createQuestionTest()
+        {
+            AdditionQuestion question = new AdditionQuestion("easy", "circle");
+            question.setProblemAndAnswer();
```
Comment (nagarkar) [inline Tests/AdditionQuestionTest.cs:16]: wtf?

File: Tests/AdditionQuestionTest.cs [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+            string problem = question.Problem;
+
+            int blankIndex = problem.IndexOf(" ");
+            int secondBlankIndex = problem.IndexOf("+", blankIndex); 
+
+            string num1 = problem.Substring(0, blankIndex + 1);
+            string num2 = problem.Substring(secondBlankIndex + 1);
```
Comment (nagarkar) [inline Tests/AdditionQuestionTest.cs:23]: This is really garbage code. why would you test something this convoluted way? Find a better way to make this class testable. Identify one obvious reason why this test is flaky and will break when the problem text format changes slightly.

File: Tests/AdditionQuestionTest.cs [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+using UnityEngine;
+using UnityEngine.TestTools;
+using Questions;
+
+namespace Tests
+{
+    public class AdditionQuestionTest
+    {
+        [Test]
+        public void createQuestionTest()
```
Comment (nagarkar) [inline Tests/AdditionQuestionTest.cs:13]: Tests should be named as :
testXYZ, where XYZ is what you want to test, 

e.g. testAdditionLogic() or testPossibleAnswersHaveNoDuplicates()

File: Questions/AdditionQuestion.cs (hunk 1 above)
Comment (aryanagarkar) [reply Questions/AdditionQuestion.cs:7]: Changed it to "Addition".

File: Questions/Question.cs (hunk 2 above)
Comment (aryanagarkar) [reply Questions/Question.cs:7]: Deleted "MonoBehavior"

File: Questions/Question.cs (hunk 4 above)
Comment (aryanagarkar) [reply Questions/Question.cs:12]: I changed the return type to a string and made the necessary changes in all of the other code.

File: Questions/Question.cs (hunk 5 above)
Comment (aryanagarkar) [reply Questions/Question.cs:32]: This is here so that the displayRenderer can put the incorrect answers on two of the circles along with the correct answer on one circle.

File: Tests/AdditionQuestionTest.cs (hunk 10 above)
Comment (aryanagarkar) [reply Tests/AdditionQuestionTest.cs:13]: Changed the name of the test to "testCreateQuestion"

File: Questions/Question.cs (hunk 4 above)
Comment (aryanagarkar) [reply Questions/Question.cs:12]: Changed this to a string and made all the necessary code changes

File: Questions/Question.cs (hunk 5 above)
Comment (aryanagarkar) [reply Questions/Question.cs:32]: Made "PossibleAnswers" an abstract property and implemented it in child classes.

File: Questions/Question.cs (hunk 3 above)
Comment (aryanagarkar) [reply Questions/Question.cs:9]: Changed difficulty, preferred display type and possible answers to read-only and made the changes.

File: Questions/Question.cs (hunk 6 above)
Comment (aryanagarkar) [reply Questions/Question.cs:46]: I took out setProblemAndAnswer() and instead called createProblem() in the constructor.

File: Tests/AdditionQuestionTest.cs (hunk 8 above)
Comment (aryanagarkar) [reply Tests/AdditionQuestionTest.cs:16]: I removed this and took out setProblemAndAnswer() and instead called createProblem() in the constructor of the question class.

File: Tests/AdditionQuestionTest.cs (hunk 9 above)
Comment (aryanagarkar) [reply Tests/AdditionQuestionTest.cs:23]: This test is flaky because if the question is in the format "num1+num2", there are no spaces so the test won't work. I added some code to the question class to make it testable. One of my properties is public, but that is for testing purposes. In the actual game, it will be protected.

File: Questions/SubtractionQuestion.cs (hunk 7 above)
Comment (aryanagarkar) [reply Questions/SubtractionQuestion.cs:15]: I made the question and answer read only, made the properties question and answer abstract, and set the questions and answers in those instead of the separate createQuestion() method.

File: State/State.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+[System.Serializable]
+public class State
+{
+    Grade grade;
+    Journey journey;
+    Topic topic;
+    int difficulty;
+
+    public State(Grade grade, Journey journey, Topic topic, int difficulty){
```
Comment (nagarkar) [inline State/State.cs:14]: Without displaytype, we can't know what question to show.

File: State/State.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        this.grade = grade;
+        this.journey = journey;
+        this.topic = topic;
+        this.difficulty = difficulty;
+    }
+
+    public Grade Grade{
+        get{return grade; }
+        set{value = grade; }
```
Comment (nagarkar) [inline State/State.cs:23]: Why is the state mutable?

File: State/StateMachine.cs [hunk 13]
```diff
@@ -0,0 +1,88 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using Questions;
+using events;
+
+public class StateMachine
+{
+    State state;
```
Comment (nagarkar) [inline State/StateMachine.cs:9]: Why state if you already have history? Shouldn't current state be top of the history?

File: State/StateMachine.cs [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    History hist;
+
+    public StateMachine()
```
Comment (nagarkar) [inline State/StateMachine.cs:12]: Shouldn't this be initialized with history?

File: State/StateMachine.cs [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+        state = new State(grade1, journey1, topic1, 5);
+        hist = new History();
+        hist.push(state);
+    }
+
+    public State CurrentState
+    {
+        get { return state; }
+        set { state = value; }
```
Comment (nagarkar) [inline State/StateMachine.cs:32]: Why is this mutable?

File: State/StateMachine.cs [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+            return getNextLevelForRightAnswer();
+        }
+        if (ev == EventTypes.wrongAnswerEvent)
+        {
+            return getNextLevelForWrongAnswer();
+        }
+        return null;
+    }
+
+    private State getNextLevelForRightAnswer()
```
Comment (nagarkar) [inline State/StateMachine.cs:53]: wrong method name?

File: State/StateMachine.cs [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+        get { return state; }
+        set { state = value; }
+    }
+
+    public History CurrentHistory{
+        get { return hist; }
+        set { hist = value; }
+    }
+
+    public State getNextState(EventTypes ev)
```
Comment (nagarkar) [inline State/StateMachine.cs:40]: why doesn't this method update history if it's doing a state transition?

Comment (nagarkar) [inline State/StateMachine.cs:40]: processEvent

File: State/StateMachine.cs [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+            hist.push(newState);
+            return newState;
+        }
+        else
+        {
+            return state;
+        }
+    }
+
+    private State getNextLevelForWrongAnswer()
```
Comment (nagarkar) [inline State/StateMachine.cs:71]: wrong method name.

File: State/StateMachine.cs [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+        {
+            State newState = new State(state.Grade, state.Journey, state.Topic, difficulty - 1);
+            state = newState;
+            hist.push(newState);
+            return newState;
+        }
+        else
+        {
+            return state;
+        }
```
Comment (nagarkar) [inline State/StateMachine.cs:86]: Avoid multiple return statements.

File: Tests/AdditionTest.cs [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        [Test]
+        public void testProblemOperandsAreReasonable()
+        {
+            for(int i = 0; i < 100; i++){
+                Addition question = new Addition(1);
+                string[] operands = question.Operands;
+
+                for(int j = 0; j < operands.Length - 1; j++){
+                    Assert.IsTrue(int.Parse(operands[j]) < 20);
```
Comment (nagarkar) [inline Tests/AdditionTest.cs:22]: very weak test.

File: Tests/StateMachineTest.cs [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+{
+    public class StateMachineTest
+    {
+        [Test]
+        public void testGetNextStateForRightAnswer()
+        {
+            StateMachine stateMachine = new StateMachine();
+
+            ArrayList topics = new ArrayList();
+            Topic topic1 = new Topic(5, "addition");
```
Comment (nagarkar) [inline Tests/StateMachineTest.cs:18]: don't use numeric suffixes when there is only one number.

File: Tests/StateMachineTest.cs [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+            ArrayList journeys = new ArrayList();
+            Journey journey1 = new Journey(topics);
+            journeys.Add(journey1);
+            Grade grade1 = new Grade(journeys);
+
+            State state;
+            State actualState;
+            State expectedState;
+
+            state = new State(grade1, journey1, topic1, 2);
```
Comment (nagarkar) [inline Tests/StateMachineTest.cs:29]: This is testDifficultIncreasesForLowDifficultyLevels_rightAnswer().

File: Tests/StateMachineTest.cs [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+            stateMachine.CurrentState = state;
+            actualState = stateMachine.getNextState(EventTypes.rightAnswerEvent);
+            expectedState = new State(grade1, journey1, topic1, 3);
+            Assert.IsTrue(actualState.isEqual(expectedState));
+
+            state = new State(grade1, journey1, topic1, 5);
+            stateMachine.CurrentState = state;
+            actualState = stateMachine.getNextState(EventTypes.rightAnswerEvent);
+            expectedState = new State(grade1, journey1, topic1, 5);
```
Comment (nagarkar) [inline Tests/StateMachineTest.cs:38]: This is testNoChangeInDifficultAtMaxDifficulty_singleTopic_rightAnswer()

File: Tests/StateMachineTest.cs [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            StateMachine stateMachine = new StateMachine();
+
+            ArrayList topics = new ArrayList();
+            Topic topic1 = new Topic(5, "addition");
+            topics.Add(topic1);
+            ArrayList journeys = new ArrayList();
+            Journey journey1 = new Journey(topics);
+            journeys.Add(journey1);
+            Grade grade1 = new Grade(journeys);
```
Comment (nagarkar) [inline Tests/StateMachineTest.cs:23]: All the code in this section is repeated in each test. Write a test helper method
CreateStateFor(string questionType, int difficulty)

File: Questions/Addition.cs [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+
+            problem = string.Join(" + ", operands);
+            answer = (int.Parse(operands[0]) + int.Parse(operands[1])).ToString();
+
+            wrongAnswers = new string[2];
+            int error = 2;
+            wrongAnswers[0] = (int.Parse(answer) - error).ToString();
+            wrongAnswers[1] = (int.Parse(answer) + error).ToString();
+
+            preferredDisplayType = "CircleDisplay";
```
[the remark this answers is not in the document]
Comment (aryanagarkar) [reply Questions/Addition.cs:21]: Created a preferredDisplayType enum and changed this to an enum type.

File: Questions/Addition.cs [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+        {
+            int len = getNumOperandsForDifficulty(difficulty);
+            string[] operands = new string[len];
+
+            if (difficulty.Equals(1))
+            {
+                for (int i = 0; i < len; i++)
+                {
+                    operands[i] = Random.Range(1, 10).ToString();
+                }
```
[the remark this answers is not in the document]
Comment (aryanagarkar) [reply Questions/Addition.cs:34]: Change this to use switch case.

File: Questions/Question.cs [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+
+        public string[] DisplayTypes
+        {
+            get
+            {
+                return displayTypes;
+            }
+        }
+
+        public string PreferredDisplayType
```
[the remark this answers is not in the document]
Comment (aryanagarkar) [reply Questions/Question.cs:54]: Created a preferredDisplayType enum and changed this to an enum type.

File: State/State.cs (hunk 11 above)
Comment (aryanagarkar) [reply State/State.cs:14]: Added a display type to state and made the necessary changes.

File: State/State.cs (hunk 12 above)
Comment (aryanagarkar) [reply State/State.cs:23]: God rid of the set accessors.

File: State/StateMachine.cs (hunk 13 above)
Comment (aryanagarkar) [reply State/StateMachine.cs:9]: Took current state out and instead used hist.peek() wherever I was getting current state.

File: State/StateMachine.cs (hunk 16 above)
Comment (aryanagarkar) [reply State/StateMachine.cs:53]: Changed "getNextLevelForRightAnswer()" to "getNextStateForRightAnswer()" and "getNextLevelForWrongAnswer()" to "getNextStateForWrongAnswer()"

File: State/StateMachine.cs (hunk 17 above)
Comment (aryanagarkar) [reply State/StateMachine.cs:40]: Changed this method to update history.

Comment (aryanagarkar) [reply State/StateMachine.cs:40]: Changed to "processEvent()".

File: State/StateMachine.cs (hunk 18 above)
Comment (aryanagarkar) [reply State/StateMachine.cs:71]: Changed "getNextLevelForWrongAnswer()" to "getNextStateForWrongAnswer()"

File: State/StateMachine.cs (hunk 19 above)
Comment (aryanagarkar) [reply State/StateMachine.cs:86]: Changed to include only one return statement.

File: Tests/StateMachineTest.cs (hunk 21 above)
Comment (aryanagarkar) [reply Tests/StateMachineTest.cs:18]: Took out numeric suffixes.

File: State/StateMachine.cs (hunk 14 above)
Comment (aryanagarkar) [reply State/StateMachine.cs:12]: Changed the constructor to include a history parameter and made all the necessary changes.

File: State/StateMachine.cs (hunk 15 above)
Comment (aryanagarkar) [reply State/StateMachine.cs:32]: I took this out, look at your earlier comment about this. I made history immutable too.

File: State/State.cs (hunk 12 above)
Comment (aryanagarkar) [reply State/State.cs:23]: "got", not "god"

File: Tests/StateMachineTest.cs (hunk 24 above)
Comment (aryanagarkar) [reply Tests/StateMachineTest.cs:23]: Wrote the helper method and replaced the repeated code.

File: Tests/StateMachineTest.cs (hunk 23 above)
Comment (aryanagarkar) [reply Tests/StateMachineTest.cs:38]: Changed this name and created separate tests.

File: Tests/StateMachineTest.cs (hunk 22 above)
Comment (aryanagarkar) [reply Tests/StateMachineTest.cs:29]: Changed this name and created separate tests.

File: Tests/AdditionTest.cs (hunk 20 above)
Comment (aryanagarkar) [reply Tests/AdditionTest.cs:22]: Changed this test
